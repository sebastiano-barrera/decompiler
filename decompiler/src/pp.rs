use std::io::Write;

use smallvec::SmallVec;

pub struct PrettyPrinter<W> {
    wrt: W,
    indent_stack: Vec<u16>,
    cur_text: u16,
    cur_indent: u16,
    indent_next_chunk: bool,
}

pub trait PP: Write {
    fn open_box(&mut self);
    fn close_box(&mut self);
}

impl<W: Write> PrettyPrinter<W> {
    pub fn start(wrt: W) -> Self {
        PrettyPrinter {
            wrt,
            indent_stack: Vec::new(),
            cur_text: 0,
            cur_indent: 0,
            indent_next_chunk: false,
        }
    }

    pub(crate) fn writer_mut(&mut self) -> &mut W {
        &mut self.wrt
    }
}

impl<W: Write> PP for PrettyPrinter<W> {
    fn open_box(&mut self) {
        self.indent_stack.push(self.cur_indent);
        self.cur_indent += self.cur_text;
        self.cur_text = 0;
    }

    fn close_box(&mut self) {
        let prev_indent = self.indent_stack.pop().unwrap();
        if !self.indent_next_chunk {
            self.cur_text += self.cur_indent - prev_indent;
        }
        self.cur_indent = prev_indent;
    }
}

impl<W: Write> Write for PrettyPrinter<W> {
    fn write(&mut self, s: &[u8]) -> std::io::Result<usize> {
        let mut cur_text = self.cur_text as usize;

        for line in s.split_inclusive(|ch| *ch == b'\n') {
            if self.indent_next_chunk {
                for _ in 0..self.cur_indent {
                    self.wrt.write_all(b" ")?;
                }
            }

            self.wrt.write_all(line)?;
            if line.ends_with(b"\n") {
                cur_text = 0;
                self.indent_next_chunk = true;
            } else {
                cur_text += line.len();
                self.indent_next_chunk = false;
            }
        }

        self.cur_text = cur_text.try_into().unwrap();

        Ok(s.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.wrt.flush()
    }
}

pub struct IoAsFmt<W>(pub W);

impl<W: std::io::Write> std::fmt::Write for IoAsFmt<W> {
    fn write_str(&mut self, s: &str) -> std::fmt::Result {
        self.0.write_all(s.as_bytes()).unwrap();
        Ok(())
    }
}

pub struct FmtAsIoUTF8<W>(pub W);

impl<W: std::fmt::Write> std::io::Write for FmtAsIoUTF8<W> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        let s = std::str::from_utf8(buf).unwrap();
        self.0.write_str(s).unwrap();
        Ok(s.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

pub struct MultiWriter<W>(SmallVec<[W; 4]>);

impl<W> MultiWriter<W> {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        MultiWriter(SmallVec::new())
    }

    pub fn add_writer(&mut self, wrt: W) {
        self.0.push(wrt);
    }
}
impl<W: std::io::Write> std::io::Write for MultiWriter<W> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        for wrt in &mut self.0 {
            let _ = wrt.write(buf);
        }
        Ok(buf.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        for wrt in &mut self.0 {
            let _ = wrt.flush();
        }
        Ok(())
    }
}
