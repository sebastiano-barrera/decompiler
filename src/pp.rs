use std::fmt::Write;

pub struct PrettyPrinter<W> {
    wrt: W,
    indent_stack: Vec<u16>,
    cur_text: u16,
    cur_indent: u16,
}

pub trait PP: Write {
    fn open_box(&mut self);
    fn close_box(&mut self);
}

impl<W: Write> PrettyPrinter<W> {
    pub fn start(wrt: W) -> Self {
        PrettyPrinter {
            wrt,
            indent_stack: Vec::new(),
            cur_text: 0,
            cur_indent: 0,
        }
    }
}

impl<W: Write> PP for PrettyPrinter<W> {
    fn open_box(&mut self) {
        self.indent_stack.push(self.cur_indent);
        self.cur_indent += self.cur_text;
        self.cur_text = 0;
    }

    fn close_box(&mut self) {
        let prev_indent = self.indent_stack.pop().unwrap_or(0);
        self.cur_text += self.cur_indent - prev_indent;
        self.cur_indent = prev_indent;
    }
}

impl<W: Write> Write for PrettyPrinter<W> {
    fn write_str(&mut self, s: &str) -> std::fmt::Result {
        let mut cur_text = self.cur_text as usize;
        for (ndx, line) in s.split('\n').enumerate() {
            if ndx == 0 {
                cur_text += line.len();
            } else {
                self.wrt.write_char('\n')?;
                // better way?
                for _ in 0..self.cur_indent {
                    self.wrt.write_char(' ')?;
                }
                cur_text = line.len();
            }
            self.wrt.write_str(line)?;
        }

        self.cur_text = cur_text.try_into().unwrap();
        Ok(())
    }
}

pub mod debug {
    use super::*;
    use std::sync::Mutex;

    static GLOBAL_PP: Mutex<Option<Box<dyn PP + Send>>> = Mutex::new(None);

    pub fn set_pp(pp: Box<dyn PP + Send>) {
        let mut global_pp = GLOBAL_PP.lock().unwrap();
        *global_pp = Some(pp);
    }

    pub fn with_pp(action: impl FnOnce(&mut dyn PP) -> std::fmt::Result) {
        let mut global_pp = GLOBAL_PP.lock().unwrap();
        if let Some(global_pp) = &mut *global_pp {
            action(global_pp.as_mut()).unwrap();
        }
    }
}
