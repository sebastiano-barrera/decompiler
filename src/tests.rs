#![cfg(test)]

mod callconv_x86_64 {
    use crate::pp::PrettyPrinter;
    use crate::{ast, mil, ssa, ty, x86_to_mil};

    use iced_x86::code_asm::CodeAssembler;
    use iced_x86::Instruction;

    #[test]
    fn struct_2p() {
        let input = {
            use iced_x86::code_asm::{eax, ptr, rsi};
            let mut asm = CodeAssembler::new(64).unwrap();

            let mut lbl_end = asm.create_label();
            let mut lbl_loop = asm.create_label();

            asm.mov(eax, 1).unwrap();
            asm.test(rsi, rsi).unwrap();
            asm.je(lbl_end).unwrap();
            asm.set_label(&mut lbl_loop).unwrap();
            asm.inc(eax).unwrap();
            asm.mov(rsi, ptr(rsi + 8)).unwrap();
            asm.test(rsi, rsi).unwrap();
            asm.jne(lbl_loop).unwrap();
            asm.set_label(&mut lbl_end).unwrap();
            asm.ret().unwrap();

            asm.take_instructions()
        };

        let mut types = ty::TypeSet::new();
        let tyid_char = types.add(ty::Type {
            name: "char".to_owned().into(),
            ty: ty::Ty::Int(ty::Int {
                size: 1,
                signed: ty::Signedness::Signed,
            }),
        });
        let tyid_char_ptr = types.add(ty::Type {
            name: "char*".to_owned().into(),
            ty: ty::Ty::Ptr(tyid_char),
        });
        let tyid_name_item = {
            let tyid = types.add(ty::Type {
                name: "name_item".to_owned().into(),
                ty: ty::Ty::Void,
            });
            let self_ptr = types.add(ty::Type {
                name: "name_item*".to_owned().into(),
                ty: ty::Ty::Ptr(tyid),
            });

            types.set(
                tyid,
                ty::Type {
                    name: "name_item".to_owned().into(),
                    ty: ty::Ty::Struct(ty::Struct {
                        size: 16,
                        members: vec![
                            ty::StructMember {
                                offset: 0,
                                name: "name".to_owned().into(),
                                tyid: tyid_char_ptr,
                            },
                            ty::StructMember {
                                offset: 8,
                                name: "next".to_owned().into(),
                                tyid: self_ptr,
                            },
                        ],
                    }),
                },
            );

            tyid
        };

        print_asm(&input);
        let prog = {
            let mut b = x86_to_mil::Builder::new();
            b.use_types(&types);
            x86_to_mil::callconv::read_func_params(&mut b, &[tyid_name_item]).unwrap();
            b.translate(input.iter().copied())
        }
        .unwrap();
        let output = finish_prog(prog, &types).unwrap();
        insta::assert_snapshot!(output);
    }

    fn finish_prog(prog: mil::Program, types: &ty::TypeSet) -> std::io::Result<String> {
        use std::io::Write;

        let mut out = Vec::new();
        writeln!(out)?;
        writeln!(out, "mil program = ")?;
        writeln!(out, "{:?}", prog)?;

        let mut prog = ssa::mil_to_ssa(ssa::ConversionParams::new(prog));
        crate::xform::canonical(&mut prog);
        ssa::eliminate_dead_code(&mut prog);
        writeln!(out)?;
        writeln!(out, "{:?}", prog)?;

        let ast = ast::Ast::new(&prog);
        writeln!(out)?;
        let mut pp = PrettyPrinter::start(&mut out);
        ast.pretty_print(&mut pp).unwrap();

        Ok(String::from_utf8(out).unwrap())
    }

    fn print_asm(instrs: &[Instruction]) {
        use iced_x86::{Formatter, IntelFormatter};
        let mut formatter = IntelFormatter::new();
        let mut instr_strbuf = String::new();
        for instr in instrs {
            print!("{:16x}: ", instr.ip());
            instr_strbuf.clear();
            formatter.format(&instr, &mut instr_strbuf);
            println!("{}", instr_strbuf);
        }
    }
}
