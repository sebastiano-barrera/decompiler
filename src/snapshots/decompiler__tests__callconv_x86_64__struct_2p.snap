---
source: src/tests.rs
expression: output
---
mil program = 
program  70 instrs
    0 r12 <- #pre:stack_bottom
    1 r2 <- #pre:CF
    2 r3 <- #pre:PF
    3 r4 <- #pre:AF
    4 r5 <- #pre:ZF
    5 r6 <- #pre:SF
    6 r7 <- #pre:TF
    7 r8 <- #pre:IF
    8 r9 <- #pre:DF
    9 r10 <- #pre:OF
   10 r11 <- #pre:RBP
   11 r12 <- #pre:RSP
   12 r13 <- #pre:RIP
   13 r14 <- #pre:RDI
   14 r15 <- #pre:RSI
   15 r16 <- #pre:RAX
   16 r17 <- #pre:RBX
   17 r18 <- #pre:RCX
   18 r19 <- #pre:RDX
   19 r20 <- #pre:R8
   20 r21 <- #pre:R9
   21 r22 <- #pre:R10
   22 r23 <- #pre:R11
   23 r24 <- #pre:R12
   24 r25 <- #pre:R13
   25 r26 <- #pre:R14
   26 r27 <- #pre:R15
   27 r28 <- #pre:arg0
   28 r14 <- sget8    r28,0
   29 r15 <- sget8    r28,8
   30 r28 <- undef
   31 r29 <- undef
   32 r30 <- undef
   33 r31 <- undef
   34 r32 <- undef
   35 r33 <- undef
   36 r34 <- undef
   37 r28 <- const4   1 (0x1)
   38 r16 <- v8.l4=   r16 ← r28
   39 r15 <- and      r15,r15  qword
   40 r6 <- sign     r15
   41 r5 <- is0      r15
   42 r30 <- l1       r15
   43 r3 <- parity   r15
   44 r2 <- const1   0 (0x0)
   45 r10 <- const1   0 (0x0)
   46 r28 <- jmp.if   r5,69
0x2:
   47 r28 <- l4       r16
   48 r28 <- addk     r28,1 dword
   49 r16 <- v8.l4=   r16 ← r28
   50 r10 <- false
   51 r2 <- false
   52 r6 <- sign     r28
   53 r5 <- is0      r28
   54 r29 <- l1       r28
   55 r3 <- parity   r29
0x0:
   56 r29 <- const8   8 (0x8)
   57 r29 <- add      r29,r15  qword
   58 r28 <- loadm8   addr:r29
   59 r15 <- get      r28
   60 r15 <- and      r15,r15  qword
   61 r6 <- sign     r15
   62 r5 <- is0      r15
   63 r30 <- l1       r15
   64 r3 <- parity   r15
   65 r2 <- const1   0 (0x0)
   66 r10 <- const1   0 (0x0)
   67 r28 <- not      r5
   68 r29 <- jmp.if   r28,47
0x1:
   69 r28 <- ret      r16


ssa program  18 instrs
.B0:    ;;  addr:0x0; 47 insn 0 phis
  (  2)  r15 <- #pre:RAX
         r27 <- #pre:arg0
  (  2)  r29 <- sget8    r27,8
  (  2)  r37 <- const4   1 (0x1)
  (  3)  r38 <- v8.l4=   r15 ← r37
         r41 <- is0      r29
         r46 <- jmp.if   r41,69
.B1:    ;;  preds:B0,B1  addr:0x2; 22 insn 2 phis
                  ɸ  B0     B1     
           r70    <- r29    r58    
  (  2)    r73    <- r38    r49    
         r48 <- addk     r73,1 dword
  (  3)  r49 <- v8.l4=   r73 ← r48
         r57 <- addk     r70,8 qword
  (  2)  r58 <- loadm8   addr:r57
         r62 <- is0      r58
         r67 <- not      r62
         r68 <- jmp.if   r67,47
.B2:    ;;  preds:B0,B1  addr:0x1; 1 insn 1 phis
                  ɸ  B0     B1     
           r76    <- r38    r49    
         r69 <- ret      r76


T0(): {
  let r29 = StructGet8[8](pre:arg0);
  let r38 = V8WithL4(pre:RAX, 0x1 /* 1 */);
  if IsZero(r29) {
    goto T2 (
      r76 = r38
    )
  
  }
  goto T1 (
    r70 = r29
    r73 = r38
  )
  
  T1(r70, r73): {
    let r49 = V8WithL4(r73, r73 + 1);
    if !(IsZero([r70 + 8]:8)) {
      goto T1 (
        r70 = [r70 + 8]:8
        r73 = r49
      )
    
    }
    goto T2 (
      r76 = r49
    )
  }
  
  T2(r76): {
    Ret(r76);
  }
}
